### PR TITLE
Additions and modifications to greedy_coloring and equitable_coloring

### PR DESCRIPTION
**Pull request by Rhyd-Lewis**
Summary of changes. The following files have been changed. These are considered below, with further details

**…\networkx\networkx\algorithms\coloring\equitable_colouring.py**

As before, this new code provides an equitable k-colouring for the nodes of a graph G (where k is specified by the user via the parameter “num_colors”). The original version of this code only works when the number of colours k (defined by the user) exceeds the maximum degree of G. In our new version, a heuristic algorithm is added that is used in cases where k is *less* than this value. This allows more flexibility for users. This heuristic algorithm is described and documented in my book:

Lewis, R. (2021) A Guide to Graph Colouring: Algorithms and Applications, 2nd Ed. Springer, ISBN: 978-3-030-81053-5 https://link.springer.com/book/10.1007/978-3-030-81054-2.

The implementation of this new algorithm has complexity O((n lg n) + (nk) + (m lg m)). As required, in solutions returned by this method, neighbouring vertices always receive different colours; however, the colouring is not guaranteed to be equitable.

Several new exceptions have now also been added to this code to reflect these changes and to make the existing code safer.

- NetworkXAlgorithmError: If a value of k is used that is too small (that is, the heuristic algorithm is unable to colour all nodes of G such that neighbours always have different colours
- NetworkXNotImplemente: If G is a directed graph or directed multigraph.
- ValueError: If k is not a positive integer. 
    
The documentation and comments have been modified to reflect these changes. All other parts of the code remain the same, albeit with some tidying. 

**…\networkx\networkx\algorithms\coloring\greedy_colouring.py**

Updates have also been made to the above file to improve performance and simplify the methods used for graph colouring. The main changes are:

- the deletion of the existing “interchange” operator, and the addition of a new tabu search operator
- the implementation of two high-performance constructive heuristics that are known to be exact on some topologies and produce good results on arbitrary graphs, 
- an improvement in the documentation and more information of algorithm complexity.
- Additional exceptions: in particular, these algorithms should not be used with directed graphs.

Full descriptions and analyses of these algorithms can also be found in my book:

Lewis, R. (2021) A Guide to Graph Colouring: Algorithms and Applications, 2nd Ed. Springer, ISBN: 978-3-030-81053-5 https://link.springer.com/book/10.1007/978-3-030-81054-2.

A concise description of the new tabu search algorithm is given in the comments/documentation. It is a well-regarded contemporary method that is known to produce results close to the state of the art. Here are its main benefits.

- The tabu search algorithm occupies fewer lines of code -- approximately 80 lines as opposed to interchange's 180.
- The tabu search algorithm can be used as an additional step with all the greedy heuristics currently included in the NetworkX library. In contrast, the interchange method does not work with DSatur, despite this being one of the best available greedy heuristics.
- The tabu search algorithm allows users to specify the maximum number of iterations. Better results (but longer run times) will occur with larger iteration limits.
- Compared to all existing colouring methods in the NetworkX library, the tabu search option produces solutions with fewer colours than interchange, and in less time. The following document contains further details on this: https://www.rhydlewis.eu/gcol/analysis.pdf.

The second change we have made is in the implementation if the Dsatur algorithm in NetworkX, where there is currently a slight error. In the original specification of Dsatur (given in the following paper):

Brélaz, D. (1979). "New methods to color the vertices of a graph". Communications of the ACM. 22(4), pp. 251–256.

the next vertex to be coloured should be the one with the highest saturation degree, breaking ties by selecting the vertex with the *highest degree in the subgraph induced by the uncoloured vertices*. However, the current NetworkX implementation breaks ties by only looking at the degree of the vertex in the original graph. This fault is corrected in our proposed implementation. Our new implementation also makes use of a priority queue and has a complexity of O((n+m) lg(n+m)) as opposed to the existing implementation of O(n^2). This new implementation occupies fewer lines of code and is faster for all but the densest of graphs. A further analysis of this algorithm (performance and complexity) is available in the above book.

Finally, we have also added the famous recursive largest first (RLF) algorithm as an option for the greedy colouring method. This is a well-known and high-performing method that has a complexity of O(nm). This is slightly more expensive than the existing greedy methods, although it is also known to produce better quality solutions. (Information on this can also be seen in the above book and at https://www.rhydlewis.eu/gcol/analysis.pdf.)

**…\networkx\networkx\algorithms\coloring\tests\test_coloring.py**

In the above file, several new tests have been added to ensure our new code functions as expected. We can confirm that is correct. Additional information on run times and performance can also be seen in https://www.rhydlewis.eu/gcol/analysis.pdf

### CODE REVIEW

**Review by rossbar — COMMENTED** [review]
For starters, it'd be good to ensure that any of the tests that are currently in place haven't been removed or modified. This is important in order to ensure that the changes are backward-compatible.

It's hard to tell from the diff whether there are material changes or whether these are just style issues, so the first step is resolving the style problems. A quick recipe:

```bash
pip install pre-commit
pre-commit install
pre-commit run --all-files
```

See the [contributor guide](https://networkx.org/documentation/latest/developer/contribute.html#development-workflow) for further details!

**Comment by Rhyd-Lewis** [comment]
Thank you for your advice,

The code changes I've proposed are adding some flexibility to the library,
allowing several options to be taken that were not previously allowed (e.g.
using  the "DSATUR" option to be used while "interchange=True").

This extra flexibility makes a couple of existing tests unnecessary. If I
am not allowed to remove these tests, what would be your advice here?

Sorry, but I've not done this before.

Thank you

Rhyd Lewis

**Comment by rossbar** [comment]
No worries, @Rhyd-Lewis - and thanks for the proposal!

As mentioned above, the first step will be to fix the style issues: at this point, it's not clear to reviewers how extensive the changes to the test suite actually are, as the modifications are intertwined with style/reformatting changes. Once those are rectified (see procedure in the previous comment, or [the contributor guide](https://networkx.org/documentation/latest/developer/contribute.html) it will be more straightforward to tell whether there are any backwards-incompatible changes and to what degree (ha!). 

At a higher level - the reason why this constraint exists is that these functions are already in use by users for all sorts of applications. It's important to ensure that any changes that go into the library don't silently yank the rug out from under them and change the behavior of *existing* code, at least not without some form of forewarning.

**Comment by Rhyd-Lewis** [comment]
Hello. All tests (except for reviewer label) have now been passed. Looking forward to receiving any feedback. Thanks
Rhyd

**Comment by dschult** [comment]
I took a look at the PR and I think it is changing the interface in a way I find confusing at first glance. Perhaps I am missing something -- or maybe we just need to make a new function instead of changing an existing function. Making a new function could also make this easier to review -- (it looks like almost every line has changed, so it's hard to tell what this does to t he existing function).

My questions:  
- this seems to provide a heuristic method that (sometimes) does not return an equitable coloring. That seems confusing given the name of the function.
- this seems to add a few new strategies for coloring -- but the diff makes it look like you are also changing/replacing an existing strategy.  Can you make the additions without removing any existing strategies? Even if the strategy is essentially the same as an existing one, it is probably better to add a new one and leave the old so people counting on it can continue to use it.

minor nit:  can you use `log(m)` instead of `lg m` for the documentation of complexity? push back if that's not ok.

**Comment by Rhyd-Lewis** [comment]
Hi Dan,

Thanks for your message and your questions. I appreciate you giving your
time

Answer 1) Currently the equitable coloring algorithms on NetworkX only
operate if the user specifies a number of colours that is larger than the
maximum degree of the graph. (It produces an exception otherwise). The
algorithm then produces a perfectly equitable colouring. In practice, using
a smaller number of colours leads to an NP-hard problem; however, equitable
colourings are still achievable in many cases. My code has simply added to
this -- the new heuristic is only used in the cases where the existing
algorithm would have returned an exception. I've provided a few more
details (on complexity and accuracy) in the first post above.

Answer 2) Yes, it would be possible to add my new (tabu search) strategy
without removing the existing "interchange" strategy. However, the
interchange strategy comes from a book written in the early 1980s and is
quite slow and inaccurate compared to contemporary approaches (evidence for
this is provided in the first post above). Finally, for some reason, the
interchange operator only works with some of the greedy methods listed in
this file and returns exceptions otherwise. My thoughts were that it would
be tidier to just get rid of this approach because the new tabu search
algorithm rectifies all of these problems. For backwards compatibility I
have, however, kept the "interchange" option in the API. If it is selected,
it simply calls the tabu search algorithm for n iterations. The first post
above shows how this leads to better results and run times.

So, these were my thoughts. But as this is my first time contributing to
NetworkX, I am happy to follow your recommendations. Let me know your
advice, and I will endeavour to follow it.

Rhyd Lewis

PS: FYI, the book that I have referenced in the code is my own:

Lewis, R. (2021) A Guide to Graph Colouring: Algorithms and Applications,
2nd Ed. Springer, ISBN: 978-3-030-81053-5
https://link.springer.com/book/10.1007/978-3-030-81054-2.

**Comment by Rhyd-Lewis** [comment]
Hi All,

Awaiting feedback on most recent commit.

Thanks

Rhyd

File: networkx/algorithms/coloring/greedy_coloring.py [hunk 1]
```diff
@@ -258,75 +493,118 @@ def strategy_saturation_largest_first(G, colors):
...
-def greedy_color(G, strategy="largest_first", interchange=False):
-    """Color a graph using various strategies of greedy graph coloring.
-
-    Attempts to color a graph using as few colors as possible, where no
-    neighbors of a node can have same color as the node itself. The
-    given strategy determines the order in which nodes are colored.
-
-    The strategies are described in [1]_, and smallest-last is based on
-    [2]_.
+def greedy_color(G, strategy="largest_first", tabu=0, interchange=False):
```
Comment (rossbar) [inline networkx/algorithms/coloring/greedy_coloring.py:502]: The new kwarg should be added at the end of the signature in order to prevent breaking changes in existing code which rely on positional arguments. See [the contributor guide](https://networkx.org/documentation/stable/developer/deprecations.html#policy) for further details if interested!

File: networkx/algorithms/coloring/tests/test_coloring.py [hunk 2]
```diff
@@ -22,72 +18,89 @@
...
     "DSATUR",
+    "rlf",
 ]
 
-# List of strategies where interchange=True results in an error
-INTERCHANGE_INVALID = ["independent_set", "saturation_largest_first", "DSATUR"]
-
 
 class TestColoring:
-    def test_basic_cases(self):
```
Comment (rossbar) [inline networkx/algorithms/coloring/tests/test_coloring.py:32]: This test suite is certainly hard to follow and the efforts to improve it are much appreciated! However, with the removal/reorganization of tests it becomes very difficult to evaluate whether the changes to the code are not breaking tested cases.

The *best* way to push this forward IMO would be to split out test suite *modifications* (i.e. changes and removals) to a separate PR that is dedicated to refactor the test suite. New tests can (and should) be *added* to this PR, but removing/changing existing tests should be left off until after the proposed changes can be shown to pass the existing test suite as is (plus any new tests).

**Review by rossbar — COMMENTED** [review]
I think probably the easiest path forward here to to try to decompose the changes into multiple PRs - see comments below. The changes generally look good, but they are extensive and should undergo careful review! Most of my suggestions at this stage are about making that review easier!

**Review by dschult — COMMENTED** [review]
Here are some comments on the equitable_coloring.py portion of the PR.
For the greedy coloring (as for the others too) we need to keep the old behavior -- including default behavior for any existing functions. At least through [a deprecation cycle](https://networkx.org/documentation/stable/developer/deprecations.html). I haven't worked through the implications of that because it looks like a lot has been removed that perhaps doesn't need to be.  It would be helpful if you passed through the PR with a view toward backward compatibility and made it more clear what is changed and how users get the old behavior. Introducing new functions can be part of that -- you don't have to rewrite existing functions. 

Thanks very much...

[Edit: Sorry -- I didn't see @rossbar's comments before I posted this.]

File: networkx/algorithms/coloring/equitable_coloring.py [hunk 3]
```diff
@@ -386,120 +388,195 @@ def procedure_P(V_minus, V_plus, N, H, F, C, L, excluded_colors=None):
...
-
-    G = nx.relabel_nodes(G, nodes_to_int, copy=True)
-
-    # Basic graph statistics and sanity check.
-    if len(G.nodes) > 0:
-        r_ = max(G.degree(node) for node in G.nodes)
+    """
+    if nx.is_directed(G):
+        raise nx.NetworkXNotImplemented("input graph cannot be directed.")
+    if not isinstance(num_colors, int) or num_colors < 0:
```
[the remark this answers is not in the document]
Comment (eriknw) [reply networkx/algorithms/coloring/equitable_coloring.py:506]: Here is an example in networkx that checks whether an input is integral:
https://github.com/networkx/networkx/blob/fd27fb0d9a579e6dd2e58f1f7762456141304ebb/networkx/generators/classic.py#L805
This works with numpy integers:
```python
>>> isinstance(np.int64(5), int)
False

>>> isinstance(np.int64(5), numbers.Integral)
True
```
I think this works b/c numpy registers the int type [here](https://github.com/numpy/numpy/blob/9142f32573612d544b65649cf6f485f55b48e97a/numpy/_core/numerictypes.py#L622).

NumPy uses a different method to check for integers, such as:
```python
import operator
try:
    operator.index(maybe_integer)
except TypeError:
    raise TypeError('blah blah') from None
```
NumPy uses this trick lots of places, such as here:
https://github.com/numpy/numpy/blob/9142f32573612d544b65649cf6f485f55b48e97a/numpy/lib/_npyio_impl.py#L822-L828
This works b/c integral objects are those that implement `__index__`.

**Comment by Rhyd-Lewis** [comment]
Thanks for the comments. I'm on leave for the next week, but will step to
it as soon as I'm back. Thanks! Rhyd

**Comment by Rhyd-Lewis** [comment]
Thank you for all your comments. As requested, I have made updated the three files so that they

(a)	are as similar as possible to the current codebase, making a review easier
(b)	take all previous comments into account.

I’ve just a few points to note.

•	The new test file is more-or-less the same as the existing one. I have added some further tests and edited one (test_strategy_saturation_largest_first). The reason for the edit is that this test seems to apply for cases where some of the vertices are already coloured (and must be ignored by saturation_largest_first). However, this option is not available in greedy_color(…), which assumes that none of the nodes are colored, nor is it permitted with any of the other strategies. It therefore seems safe to remove.
•	In summary, one heuristic algorithm has been added to equitable_coloring.py, and three have been added to greedy_coloring.py. The docstrings have been edited to reflect this, and all existing behaviour is maintained.
•	I would recommend that the interchange operator is removed in future. (A deprecation warning is needed). This is due to the performance issues documented above.

**Comment by Rhyd-Lewis** [comment]
Thank you for all your comments. As requested, I have pushed new versions that 

(a)	are as similar as possible to the current codebase, making a review easier
(b)	take all previous comments into account.

I’ve just a few points to note.

•	The test suite is more-or-less the same as before. I have added some further tests and edited one (test_strategy_saturation_largest_first). The reason for the edit is that this test seems to apply for cases where some of the vertices are already coloured (and must be ignored by saturation_largest_first). However, this option is not available in greedy_color(…), which assumes that none of the nodes are colored, nor is it permitted with any of the other strategies. It therefore seems safe to remove.
•	In summary, one heuristic algorithm has been added to equitable_coloring.py, and three have been added to greedy_coloring.py. The docstrings have been edited to reflect this, and all existing behaviour is maintained.
•	I would recommend that the interchange operator is removed in future. (A deprecation warning is needed). This is due to the performance issues documented above.

**Comment by Rhyd-Lewis** [comment]
Thank you for all your comments. As requested, I have pushed new versions that (a) are as similar as possible to the current codebase, making a review easier (b) take all previous comments into account.

I’ve just a few points to note.

The test suite is more-or-less the same as before. I have added some further tests and edited one (test_strategy_saturation_largest_first). The reason for the edit is that this test seems to apply for cases where some of the vertices are already coloured (and must be ignored by saturation_largest_first). However, this option is not available in greedy_color(…), which assumes that none of the nodes are colored, nor is it permitted with any of the other strategies. It therefore seems safe to remove.

In summary, one heuristic algorithm has been added to equitable_coloring.py, and three have been added to greedy_coloring.py. The docstrings have been edited to reflect this, and all existing behaviour is maintained.

I would recommend that the interchange operator is removed in future. (A deprecation warning is needed). This is due to the performance issues documented above.

**Comment by Rhyd-Lewis** [comment]
Hi,

Would someone be able to take a look at the new code submission? I put it up a couple of weeks ago.

Thanks

Rhyd Lewis

**Comment by Rhyd-Lewis** [comment]
Hi Ross,

Thanks for your recent feedback on my NetworkX code (see below).

I updated things according to the feedback I received a few months ago, but
have not received any responses since.

Would you be able to have a look for me?

Thanks

Rhyd Lewis

**Comment by Rhyd-Lewis** [comment]
Hi @rossbar @dschult @eriknw. All suggeted corrections made above have been made. These were made some time ago and I've been awaiting a response since the 12th of June. Thanks, @Rhyd-Lewis

**Review by rossbar — COMMENTED** [review]
Apologies @Rhyd-Lewis - the problem for me is that I still can't see the forest through the trees. I think the only way forward (for me at least) is to decompose the changes into multiple PRs where each one has a well-defined boundary per feature proposal. AFAICT, this PR includes:
 - The addition of a heuristic for certain cases in equitable coloring
 - Changes to the saturation largest first strategy
 - Introduction of a new rlf strategy
 - Introduction of a new postprocessing function to try to reduce the number of colors found during typical running of greedy coloring
